### ISSUE TITLE
Parallel Ask-and-Tell

### ISSUE THREAD

**Issue by bacalfa** [facebookresearch/nevergrad#709]
Is it possible to use the ask-and-tell framework in parallel? I can't find examples. I have something as follows (not the complete code):

```python
def ask_and_tell(optimizer, func):
    x = optimizer.ask()
    loss = func(*x.args)
    optimizer.tell(x, loss)

    return x, loss


def run_Optimizer(optimizer, func, max_time=None):
    with multiprocessing.Pool(processes=multiprocessing.cpu_count() - 1) as pool:
        all_args = [(optimizer, func) for _ in range(optimizer.budget)]
        results = sorted(pool.starmap_async(ask_and_tell, all_args).get(max_time), key=lambda r: r[1])
        best_x = results[0][0]
        best_loss = results[0][1]

    return {"x": best_x, "fun": best_loss}
```

But I'm getting an error saying a local object can't be pickled. Am I doing it right? Or is the problem somewhere in my code?

**Comment by bacalfa** [comment]
Hold on. I think I know what the issue is. But I'd appreciate some comments on how I used `multiprocessing` and the ask-and-tell framework. :)

**Comment by bacalfa** [comment]
It's not right. I'm creating copies of the `optimizer` object this way. It should be the same object across all parallel processes.

**Comment by bacalfa** [comment]
I think I got it to work by changing the approach. The basic idea is to "manually" call `ask` a few times, evaluate the black-box in parallel, then call `tell` for the corresponding `(x, loss)` tuples. Comments and suggestions are welcome!

```python
best_x = None
best_loss = np.inf
t_start = timeit.default_timer()
while True:
    num_workers = max(0, min(optimizer.num_workers, optimizer.budget - optimizer.num_tell))
    if num_workers > 0:
        curr_asks = [optimizer.ask() for _ in range(num_workers)]

        with multiprocessing.Pool(processes=num_workers) as pool:
            all_args = [(x,) for x in curr_asks]
            results = sorted(pool.starmap_async(func, all_args).get(max_time), key=lambda r: r[1])
            [optimizer.tell(result[0], result[1]) for result in results]

            if results[0][1] < best_loss:
                best_x = list(results[0][0].args)
                best_loss = results[0][1]

    if optimizer.num_tell >= optimizer.budget:
        break

    if max_time is not None:
        if timeit.default_timer() - t_start >= max_time:
            break
```

In the above, `func` represents the black-box simulator that returns the `x` parametrization point and its associated `loss` value. Also, `max_time` is the budget in terms of maximum time.

**Comment by jrapin** [comment]
Yes you do need to have a "master" process where the optimizer is running. With the your initial multiprocessing code, the optimizers where duplicated and therefore became completely independent.
You can see how `minimize` implements the multiprocesing functionality using an executor:
https://github.com/facebookresearch/nevergrad/blob/b873188e33f935531004a0f3d72a8978c39f3dfc/nevergrad/optimization/base.py#L453-L494
Why don't you use it directly by the way? https://facebookresearch.github.io/nevergrad/optimization.html#using-several-workers
It could save you a lot of hassle

**Comment by bacalfa** [comment]
Yes, I found about it later. I wanted to have more control over the optimization, that's why I opted for the ask-and-tell framework.

**Comment by bacalfa** [comment]
Also, does `ThreadPoolExecutor` offer real parallelism with CPython? See my StackOverflow post [here](https://stackoverflow.com/questions/61820762/multi-threading-availability-in-python).

**Comment by jrapin** [comment]
> Yes, I found about it later. I wanted to have more control over the optimization, that's why I opted for the ask-and-tell framework.

The other option to have *some* control is to use minimize but adding [callbacks](https://facebookresearch.github.io/nevergrad/optimization.html#adding-callbacks), which let you check what is happening (but that's only for checking, not *interacting*)

> Also, does ThreadPoolExecutor offer real parallelism with CPython?

I doubt it, except if you have IO operations. In my (little) understanding there are nearly the only ones that parallelize well over threads in Python. Can you use ProcessPoolExecutor instead? It *may* offer better parallelization. But that does not always convert to an actual speed up :s

**Comment by bacalfa** [comment]
I see. But how do I handle my other [question](https://github.com/facebookresearch/nevergrad/issues/703) regarding the best `loss` value? If I call `optimizer.minimize`, I only get back the recommendation, but not its associated loss. Or am I missing something?

**Comment by jrapin** [comment]
That's what the callback can do for you: you create a callable object that keeps track of the best loss so far and best parameter so far, and register it as a callback for the `tell` method

**Comment by bacalfa** [comment]
I'll give it a try and let you know. Is it possible to specify a time budget? I couldn't find a way to do that. Only in terms of maximum number of iterations. If there isn't, I'd suggest this as a new feature. :)

**Comment by bacalfa** [comment]
Another suggestion. [Here](https://github.com/facebookresearch/nevergrad/blob/master/nevergrad/optimization/base.py#L446), it assumes there haven't been any `tell`s yet. I'm evaluating an initial guess with ask-and-tell, and then calling `minimize`. So the `budget` has to reflect that. Or let me know if there's another way of accomplishing the same.

**Comment by jrapin** [comment]
>  Is it possible to specify a time budget?

Not yet, for now the only budget available is a number of `tell`. There is a longer term goal for this, I just created an issue to track it #714 

> it assumes there haven't been any tell yet

I don't thinks it matters at all actually, looking at where it is used, it's only there to initialize the timer which waits for results. If I remember correctly it basically makes sure we don't `sleep` in the first iteration and then adapt the sleeping time to the time it takes to evaluate the function. So no influence on whether you had previous `tell`s or not.

**Comment by teytaud** [comment]
Presumably this issue is solved and can be closed. Ok for you ?
